classes/local/checklist_comment_student.php

```diff
@@ -0,0 +1,146 @@
+            ),
+            'text' => array(
+                'type' => PARAM_TEXT,
+            ),
+        );
+    }
+
+    /**
+     * Get all matching comments by user id and item ids
+     * @param int $userid
+     * @param int[] $itemids
+     * @return checklist_comment_student[] $itemid => $check
+     */
+    public static function get_student_comments_indexed($userid, $itemids): array {
+        global $DB;
+
+        $ret = [];
+        if (!$itemids) {
+            return $ret;
+        }
+
+        list($isql, $params) = $DB->get_in_or_equal($itemids, SQL_PARAMS_NAMED);
+        $params['usermodified'] = $userid;
+        $studentcomments = self::get_records_select("usermodified = :usermodified AND itemid $isql", $params);
+
+        foreach ($studentcomments as $comment) {
+            $ret[$comment->get('itemid')] = $comment;
+        }
+        return $ret;
+    }
+
+    /**
+     * Get the name of the person who made the comment
+     * @return string|null
+     */
+    public function get_commentby_name() {
+        return $this->studentname;
+    }
+
+    /**
+     * Add the name of the commenter to the given comments.
+     * @param checklist_comment_student[] $studentcomments
+     */
+    public static function add_student_names($studentcomments) {
+        global $DB;
+
+        $userids = [];
+        foreach ($studentcomments as $studentcomment) {
+            if ($studentcomment->get('usermodified')) {
+                $userids[] = $studentcomment->get('usermodified');
+            }
+        }
+        if (!$userids) {
+            return;
+        }
+
+        if (class_exists('\core_user\fields')) {
+            $namesql = \core_user\fields::for_name()->get_sql('', true);
+        } else {
+            $namesql = (object)[
+                'selects' => ','.get_all_user_name_fields(true),
+                'joins' => '',
+                'params' => [],
+                'mappings' => [],
+            ];
+        }
+        $studentcommentusers = $DB->get_records_list('user', 'id', $userids, '', 'id'.$namesql->selects);
+        foreach ($studentcomments as $studentcomment) {
+            if ($studentcomment->get('usermodified')) {
+                if (isset($studentcommentusers[$studentcomment->get('usermodified')])) {
+                    $studentcomment->studentname = fullname($studentcommentusers[$studentcomment->get('usermodified')]);
+                }
+            }
+        }
+    }
+
+    /** Update or create a comment for a student on the given checklist item.
+     * @param int $checklistitemid id of the item in the checklist.
+     * @param string $commenttext text of the comment made by the student.
+     * @param checklist_comment_student|bool $existingcomment the comment to update or false to create a new comment record.
+     * @return bool true if successful.
+     */
+    public static function update_or_create_student_comment($checklistitemid, $commenttext, $existingcomment): bool {
```

Comment:
Comment that applies to a lot of functions, but I'll mention it specifically here: ideally all new functions should declare parameter types. I am aware that most of the existing functions do not do this, but many of them were written long before scalar parameter declarations were possible. This is not a blocker for accepting the patch, but it would be nice :-)

Reply:
Have added types to all function parameters now